[14/N] Refactor _new_process_group_helper() to remove repeated code

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88330 Allow Process Group to support multiple backends and move PG specfic implementations to backend class
* #87088 Remove ProcessGroupRoundRobin
* **#88351 [14/N] Refactor _new_process_group_helper() to remove repeated code**
* #86409 [13/N] Update gather with CPU/CUDA implementations
* #86408 [12/N] Update scatter with CPU/CUDA implementations

### Changes
- refactor parts of `_new_process_group_helper()` to remove repeated code

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D41188274](https://our.internmc.facebook.com/intern/diff/D41188274)

H-Huang committed:
[14/N] Refactor _new_process_group_helper() to remove repeated code

[ghstack-poisoned]

H-Huang committed:
Update on "[14/N] Refactor _new_process_group_helper() to remove repeated code"


Changes:
- refactor parts of `_new_process_group_helper()` to remove repeated code

[ghstack-poisoned]

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -968,9 +903,32 @@ def _new_process_group_helper(
+        if backend in [Backend.GLOO, Backend.NCCL, Backend.UCC]:
+            # In debug mode and if GLOO is available, wrap in a wrapper PG that
+            # enables enhanced collective checking for debugability.
+            if get_debug_level() == DebugLevel.DETAIL:
```
This `if` condition should be put out-most.

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -968,9 +903,32 @@ def _new_process_group_helper(
+                                GLOO is not available. Build with Gloo to
+                                create a wrapper process group in debug mode
+                                to aid collective desynchronization debugging."""
+                    )
```
Move the current out-most condition as an `elif` here:
```
elif backend not in [Backend.GLOO, Backend.NCCL, Backend.UCC]:
  logger.info(
    """TORCH_DISTRIBUTED_DEBUG was set to DETAIL, but the current backend does not support this feature."""
  )
else:
```

H-Huang replied:
Did we want to add a log for this? We currently don't have a message. I am also not sure if there are things other than `ProcessGroupWrapper` that is available when `dist.DebugLevel.DETAIL` is set, maybe the logs are also more verbose so maybe we should just say the current backend does not support `ProcessGroupWrapper`?

kwen2501 reviewed (approved):
LGTM. Thanks for refactoring.
Just added minor comments.

H-Huang committed:
Update on "[14/N] Refactor _new_process_group_helper() to remove repeated code"


Changes:
- refactor parts of `_new_process_group_helper()` to remove repeated code

[ghstack-poisoned]

awgu commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1047,9 +982,32 @@ def _new_process_group_helper(
+    # Process group wrapper initialization for supported PGs when TORCH_DISTRIBUTED_DEBUG is set
+    if backend in [Backend.GLOO, Backend.NCCL, Backend.UCC]:
+        # In debug mode and if GLOO is available, wrap in a wrapper PG that
+        # enables enhanced collective checking for debugability.
```
nit: 
```suggestion
        # enables enhanced collective checking for debuggability.
```
https://wikidiff.com/debugability/debuggability

H-Huang replied:
thank you sir

awgu reviewed (commented):
Pushing commits as well I see...

H-Huang committed:
Update on "[14/N] Refactor _new_process_group_helper() to remove repeated code"


Changes:
- refactor parts of `_new_process_group_helper()` to remove repeated code

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41188274).

H-Huang commented:
@pytorchbot merge

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41188274).
